File: ruby/affine-cipher/affine_cipher.rb
class Affine
  ALPHABET = ('a'..'z').to_a
  ALPHABET_LEN = 26

  def initialize(a, b)
    raise ArgumentError if a.gcd(ALPHABET_LEN) != 1
    @a = a
    @b = b
    @mmi = mmi
  end

  def encode(plaintext)
    plaintext.downcase
             .scan(/\w/)
             .map(&method(:encode_char))
             .each_slice(5)
             .map(&:join)
             .join(' ')
  end

  def decode(ciphertext)
    ciphertext.scan(/\w/)
              .map(&method(:decode_char))
              .join
  end


  private

  def encode_char(letter)
    x = ALPHABET.index(letter)
    return letter if x.nil?
    y = (@a*x + @b) % ALPHABET_LEN
    ALPHABET[y]
  end

  def decode_char(letter)
    y = ALPHABET.index(letter)
    return letter if y.nil?
    x = @mmi * (y - @b) % ALPHABET_LEN
    ALPHABET[x]
  end

  def mmi
    (1..ALPHABET_LEN).detect { |n| @a * n % ALPHABET_LEN == 1 }
  end

end

File: python/affine-cipher/affine_cipher.py
from string import ascii_lowercase
import math, re

ALPHABET = list(ascii_lowercase)
ALPHABET_LEN = 26

def valid_keys(func):
    def decorator(*args):
        _, a, b = args
        if math.gcd(a, ALPHABET_LEN) != 1:
            raise ValueError(f'invalid key a: {a}')
        return func(*args)
    return decorator



@valid_keys
def encode(plain_text, a, b):
    def encode_char(char):
        if char.isalpha():
            x = ALPHABET.index(char)
        else:
            return char 
        y = (a * x + b) % ALPHABET_LEN
        return ALPHABET[y]
    
    encoded_chars = [encode_char(s) for s in re.findall(r'\w', plain_text.lower())]
    encoded_chars_len = len(encoded_chars)
    return ' '.join([''.join(encoded_chars[i:(min(encoded_chars_len, i+5))])
                             for i in range(0, encoded_chars_len, 5)])

@valid_keys
def decode(ciphered_text, a, b):
    mmi = next(n for n in range(1, ALPHABET_LEN+1) if a * n % ALPHABET_LEN == 1)
    def decode_char(char):
        try:
            y = ALPHABET.index(char)
        except ValueError:
            return char
        x = mmi * (y - b) % ALPHABET_LEN
        return ALPHABET[x]
    return ''.join([decode_char(s) for s in re.findall(r'\w', ciphered_text)])